### PR TITLE
[python-package] Allow to pass Arrow array & Polars series as position

### PR DESCRIPTION
**Pull request by maxzw**
### Summary

All other metadata fields (`label`, `weight`, `group`, `init_score`) accept PyArrow ChunkedArray types as input, but `position` only accepts `numpy 1-D array` or `pandas Series`. This is inconsistent and forces users working with Arrow-native or Polars data to manually convert to NumPy before passing position data. This contributes to https://github.com/lightgbm-org/LightGBM/issues/6204.

Reference PRs (part of https://github.com/lightgbm-org/LightGBM/pull/6022)
- https://github.com/lightgbm-org/LightGBM/pull/6166
- https://github.com/lightgbm-org/LightGBM/pull/6164

### Changes

- Followed existing logic from other metadata fields to add PyArrow ChunkedArray and Polars support to the `position` field
- Updated documentation
- Added unit tests

### CODE REVIEW

**Comment by jameslamb** [comment]
Thanks for working on this! Very supportive of this change.

I haven't reviewed yet, just a quick note... CI will be blocked until someone can review this PR to fix a couple issues: #7249. Sorry for the delay.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for this, at a glance it looks pretty good to me and I support this.

There are some test failures:

```text
FAILED tests/python_package_test/test_basic.py::test_dataset_construction_overwrites_user_provided_metadata_fields - AssertionError: 
Arrays are not equal

(dtypes int32, float32 mismatch)
 ACTUAL: array([0, 1], dtype=int32)
 DESIRED: array([0., 1.], dtype=float32)
FAILED tests/python_package_test/test_engine.py::test_ranking_with_position_information_with_dataset_constructor - AssertionError: 
Arrays are not equal

(dtypes int32, int64 mismatch)
 ACTUAL: array([0, 1, 2, ..., 1, 3, 9], shape=(3005,), dtype=int32)
 DESIRED: array([0, 5, 9, ..., 5, 4, 8], shape=(3005,))
= 2 failed, 1486 passed, 15 skipped, 8 xfailed, 52 xpassed, 3395 warnings in 542.38s (0:09:02) =
```

([build link](https://github.com/lightgbm-org/LightGBM/actions/runs/25459198029/job/74696773167?pr=7260))

Can you please investigate those?

It'll be faster to run these tests locally yourself, which you can do like this:

```shell
# build lib_lightgbm (only needs to be re-run for C++ changes)
rm -rf ./build
cmake -B build -S .
cmake --build build --target _lightgbm -j4

# build and install lightgbm Python package
sh build-python.sh --precompile install

# run tests
pytest \
  tests/python_package_test/test_basic.py \
  tests/python_package_test/test_engine.py
```

Happy to review more thoroughly once the tests are passing. `@` me if you need help.

**Comment by maxzw** [comment]
> There are some test failures:
> 
> Can you please investigate those?

Oops 😄 They should be passing now!

**Comment by jameslamb** [comment]
@maxzw could you please update this to account for the changes from #7275 , and add `polars` inputs for `position`?

I'd be happy to review after that, I think we can get this into the next release.

**Comment by maxzw** [comment]
> @maxzw could you please update this to account for the changes from #7275 , and add `polars` inputs for `position`?
> 
> I'd be happy to review after that, I think we can get this into the next release.

Should be up-to-date now!

File: python-package/lightgbm/basic.py [hunk 1]
```diff
@@ -82,8 +82,10 @@
     nwt.IntoSeries,
 ]
 _LGBM_PositionType = Union[
+    List[int],
```
Comment (jameslamb) [inline python-package/lightgbm/basic.py]:
```suggestion
# 'position' intentionally does not support 'list' inputs
# ref: https://github.com/lightgbm-org/LightGBM/pull/5929#discussion_r1262646998
_LGBM_PositionType = Union[
```

It's an intentional choice to not support `list` inputs for `position`: https://github.com/lightgbm-org/LightGBM/pull/5929#discussion_r1262646998

Please revert this and anything else in this PR that allows for `position` to be a `list`.

And let's add a code comment to prevent anyone from trying to introduce this again in the future. I could imagine others seeing this as an inconsistency to be "fixed" and I don't want to rely on maintainers remembering that conversation thread.

Comment (jameslamb) [reply python-package/lightgbm/basic.py]: In general, `list` inputs for array-like data are painful to support. I'd prefer that `lightgbm` not add any more of them.

File: python-package/lightgbm/basic.py [hunk 2]
```diff
@@ -3118,8 +3120,11 @@ def set_position(
         """
         self.position = position
         if self._handle is not None and position is not None:
-            position = _list_to_1d_numpy(data=position, dtype=np.int32, name="position")
+            if isinstance(position, pd_Series) or not nwd.is_into_series(position):
+                position = _list_to_1d_numpy(data=position, dtype=np.int32, name="position")
             self.set_field("position", position)
+            # NOTE: positions are remapped to dense internal indices on the C++ side
+            # (e.g. [0, 5, 9] -> [0, 1, 2]), so we deliberately skip re-reading them here.
```
Comment (jameslamb) [inline python-package/lightgbm/basic.py]: Yes this is an inconsistency, thanks for pointing it out. I actually think it'd be a good one to resolve!

`set_{thing}` updates `self.{thing}` for all the other "info" data in `Dataset`.

https://github.com/lightgbm-org/LightGBM/blob/fa4f643c83bcdeec4b03b0dd1d2bbd55396918b2/python-package/lightgbm/basic.py#L3093-L3094

https://github.com/lightgbm-org/LightGBM/blob/fa4f643c83bcdeec4b03b0dd1d2bbd55396918b2/python-package/lightgbm/basic.py#L3063

https://github.com/lightgbm-org/LightGBM/blob/fa4f643c83bcdeec4b03b0dd1d2bbd55396918b2/python-package/lightgbm/basic.py#L3008

https://github.com/lightgbm-org/LightGBM/blob/fa4f643c83bcdeec4b03b0dd1d2bbd55396918b2/python-package/lightgbm/basic.py#L3041

I think this was just missed in #5929... I don't see any discussion about it there.

But that said, this inconsistency isn't strictly related to the data type of `position`, so that change doesn't belong here.

Are you interested in addressing this in a separate PR? It'd basically involve having `position` follow the way `weight` is handled, and adding a unit test confirming that the output of `.position` and `.get_field("position")` have the expected values, object type, and data type.

File: tests/python_package_test/test_arrow.py [hunk 3]
```diff
@@ -293,6 +293,29 @@ def test_dataset_construct_groups(group_data, arrow_type):
...
+)
+@pytest.mark.parametrize("arrow_type", _INTEGER_TYPES)
+def test_dataset_construct_position(position_data, arrow_type):
+    data = generate_dummy_arrow_table()
+    positions = pa.chunked_array(position_data, type=arrow_type)
+    dataset = lgb.Dataset(data, label=[0, 1, 0, 1, 0], position=positions, params=dummy_dataset_params())
+    dataset.construct()
+
+    expected = np.array([0, 1, 2, 3, 4], dtype=np.int32)
+    np_assert_array_equal(expected, dataset.get_field("position"), strict=True)
```
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:308]: This `test_dataset_construct_position()` test looks great.

Could you please add one more test here ensuring that cases where `position` has duplicate or out-of-order values are handled correctly? I don't see a bug in your changes, but could imagine cases where some logic bug would produce the values `0, 1, 2, 3, 4` for a 5-element `position` even when that encoding is not correct.

A test case like `position=pa.chunked_array([15, 15, 15, 27, 8, 9, 15])` or something would be a strong test of correctness. It'd also be a way to cover that difference between `get_field("position")` and `.position` that you called out in a comment in the `set_position()` body.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Implementation looks great, thank you!  Looks to me like it's closely following the patterns @borchero set up for Arrow data in previous PRs.

I just left some small suggestions about type hints and tests.

File: python-package/lightgbm/basic.py (hunk 1 above)
Comment (maxzw) [reply python-package/lightgbm/basic.py]: I've removed the support for `list` inputs 👍🏻

File: python-package/lightgbm/basic.py (hunk 2 above)
Comment (maxzw) [reply python-package/lightgbm/basic.py]: I've removed the comment regarding the remapping. I agree this change is outside the scope of this PR and should be addressed in a separate one. I'll try to find some time for it 👍🏻

File: tests/python_package_test/test_arrow.py (hunk 3 above)
Comment (maxzw) [reply tests/python_package_test/test_arrow.py:308]: Added a unit test that addresses these cases! 👍🏻

File: python-package/lightgbm/basic.py (hunk 2 above)
Comment (maxzw) [reply python-package/lightgbm/basic.py]: I created a new PR for your comment above: https://github.com/lightgbm-org/LightGBM/pull/7344

Comment (jameslamb) [reply python-package/lightgbm/basic.py]: That was fast, thanks! Just merged that.

File: tests/python_package_test/test_arrow.py [hunk 4]
```diff
@@ -293,6 +293,42 @@ def test_dataset_construct_groups(group_data, arrow_type):
...
+
+@pytest.mark.parametrize("arrow_type", _INTEGER_TYPES)
+def test_dataset_construct_position_with_duplicates_and_out_of_order(arrow_type):
+    data = generate_dummy_arrow_table()
+    positions = pa.chunked_array([[15, 15, 8, 27, 15]], type=arrow_type)
+    dataset = lgb.Dataset(data, label=[0, 1, 0, 1, 0], position=positions, params=dummy_dataset_params())
+    dataset.construct()
+
+    # positions are remapped on the C++ side to dense indices in first-seen order:
+    # 15 -> 0, 8 -> 1, 27 -> 2
```
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:319]: Love this description of how this works, thank you.

**Review by jameslamb — APPROVED** [review]
This is looking good to me, thanks!

Merging in `master` to get one more CI run with the changes from #7344 .

@borchero could you review too when you have time?

**Review by jameslamb — APPROVED** [review]
I've re-reviewed, think this is looking great. Thanks very much for this!
